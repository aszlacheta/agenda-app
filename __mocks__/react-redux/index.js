/* eslint-disable camelcase */
const react_redux = jest.genMockFromModule('react-redux');

react_redux.useDispatch = jest.fn().mockReturnValue(() => {});

module.exports = react_redux;
