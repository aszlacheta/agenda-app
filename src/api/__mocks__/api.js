const getAgenda = () => {
  return {
    then: callback => {
      const data = {
        data: {
          agendaId: '1234',
          day: 0,
          entries: []
        }
      };
      callback(data);
    }
  };
};

module.exports = {
  getAgenda
};
