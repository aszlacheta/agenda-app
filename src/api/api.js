import axios from 'axios';

export const DEFAULT_BACKEND_URL = 'http://localhost';

/**
 * Method used to get initial data of given agenda
 * @param {string} agendaId id of agenda
 * @returns
 */
export function getAgenda (agendaId) {
  return axios.get(`${DEFAULT_BACKEND_URL}/agenda/${agendaId}`)
    .catch((error) => {
      console.error(error);
    });
}

/**
 * Method used to delete entry from agenda
 *
 * @param {string} agendaId ID of the agenda
 * @param {string} entryId ID of the entry to be deleted
 * @returns {Promise}
 */
export function deleteAgendaEntry (agendaId = '', entryId = '') {
  return axios.delete(`${DEFAULT_BACKEND_URL}/agenda/${agendaId}/${entryId}`)
    .catch((error) => {
      console.error(error);
    });
}

/**
 * Method used to add new entry to given agenda
 * @param {string} agendaId
 * @param {{startDate, endDate, name, description}} entry
 * @returns {Promise}
 */
export function addAgendaEntry (agendaId = '', entry = {}) {
  return axios.post(`${DEFAULT_BACKEND_URL}/agenda/${agendaId}`, entry)
    .catch(error => {
      console.error(error);
    });
}
