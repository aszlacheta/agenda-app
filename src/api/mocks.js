import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import moment from 'moment';
import { getRandomInt } from '../components/agenda/Agenda';
import { DEFAULT_BACKEND_URL } from './api';

const mock = new MockAdapter(axios);

// GET agenda details
mock.onGet(new RegExp(`${DEFAULT_BACKEND_URL}\/agenda\/\\d+/*`))
  .reply(200, {
    id: '1234',
    day: moment('2022-07-01').valueOf(),
    entries: [
      {
        id: getRandomInt(),
        startDate: moment('2022-07-01 10:15:00').valueOf(),
        name: "Start of the training day",
        description: "Start of the training day description",
        editable: false,
        deletable: false,
      },
      {
        id: getRandomInt(),
        startDate: moment('2022-07-01 18:30:00').valueOf(),
        name: "End of the training day",
        description: "End of the training day description",
        editable: false,
        deletable: false,
      },
    ]
  });

mock.onDelete(new RegExp(`${DEFAULT_BACKEND_URL}\/agenda\/\\d+\/\\d+/*`))
  .reply(200, {
    message: 'Success',
  });

mock.onPost(new RegExp(`${DEFAULT_BACKEND_URL}\/agenda\/\\d+/*`))
  .reply(200, {
    message: 'Success',
  });

