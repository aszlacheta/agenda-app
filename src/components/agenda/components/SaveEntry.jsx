import React from 'react';
import { useDispatch, useSelector } from "react-redux"
import { addAgendaEntry } from '../../../api/api';
import { getAgendaId } from '../agendaSelectors';
import { setEntrySaved } from '../agendaSlice';

export default function SaveEntry({ entryId, form }) {
    const agendaId = useSelector(getAgendaId);
    const dispatch = useDispatch();

    const onClick = () => {
        form.startDate = form.startDate.valueOf();

        addAgendaEntry(agendaId, form)
            .then(() => {
                dispatch(setEntrySaved({ entryId, form }));
            });
    }

    return <button className="save-entry" onClick={onClick}>save</button>
}