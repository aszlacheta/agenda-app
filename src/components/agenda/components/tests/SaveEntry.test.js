import 'jsdom-global/register';
import React from 'react';
import renderer from 'react-test-renderer';
import SaveEntry from '../SaveEntry';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import moment from 'moment';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { addAgendaEntry } from '../../../../api/api';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore();
jest.mock('../../../../api/api');
jest.unmock('react-redux');

describe('SaveEntry component', () => {
  const entryId = 'testId';
  const entry = {
    startDate: moment(),
    name: 'Test name',
    description: 'Test description'
  };
  const initialState = {
    agenda: {
      id: 'test'
    }
  };
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it('should render properly', () => {
    const tree = renderer
      .create(<Provider store={store}><SaveEntry entryId={entryId} entry={entry} /></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should save agenda entry on click', () => {
    const wrapper = mount(<Provider store={store}><SaveEntry entryId={entryId} entry={entry} /></Provider>);
    const button = wrapper.find('button');

    button.simulate('click');

    expect(addAgendaEntry).toHaveBeenCalledWith(initialState.agenda.id, entry);
  });
});
