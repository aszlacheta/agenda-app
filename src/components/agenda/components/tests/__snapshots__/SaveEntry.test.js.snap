// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveEntry component should render properly 1`] = `
<button
  className="save-entry"
  data-cy="save-button"
  disabled={false}
  onClick={[Function]}
>
  entry.buttons.save, 
</button>
`;
